### DISCUSSION TITLE
"init" as a Modifier for Methods.

### DISCUSSION THREAD

**Discussion by mkwpaul** [dotnet/csharplang#4454] [Language Ideas]
Currently you can only set the value of readonly fields and auto-implemented get-only properties in the constructor of a class. Attempting to do so anywhere else will generate a compiler-error. 

Now this can cause some issues when you have a number of get-only auto properties where you can't inline the initialization. 
One example that comes to my mind are RelayCommands in ViewModels in MVVM Applications. Having the initilzation of all commands in the constructor of a ViewModel can get unwieldly, especially if there are other things that need to be done at initilzation. 

I think a way to mark a method as an "initilzation method", which could only be called from a constructor and could in turn set the values for readonly fields and auto-implemented get-only properties, would be really helpful. That way I could keep the code in my constructor clean and high-level and also more reusable if I have multiple constructors. Even just having the ability to limit a method to only being callable from a constructor would also help to make your intent clearer and to enforce that intent. 

And I think reusing the recently added "init" keyword as modifier for methods for that purpose would be just perfect. It keeps the same semantics of only being accesible at object initilzation and would be, in my opinion, very intuitive to understand. 

Now since these methods would only be callable from the classes constructor, they'd necessairly have to be either protected or private methods. A public init method would not make much sense, since you wouldn't be able to call them on an instance of that class anyway. Whether a public init method would create a warning or an error, I'm not sure, but I'm leaning more toward error.

And since constructors cannot be async, allowing init methods to be async would probably also cause more trouble that it's worth. Although I'm not sure about that one.

Thoughts?

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-06-15.md#init-methods

**Comment by pinkfloydx33** [comment]
You can already write methods that can be called from a constructor. Why not just return a value (or values) after performing your calculations and assign the return value to the properties/fields still within the constructor. That's reusable, can be called from multiple constructors and doesn't require any language changes.

If you make it `static` (which it sounds like it should be, if based on parameters only and not partially constructed state) you can also mark it `internal` to keeps its visibility down while allowing you to test if directly. I feel like an `init` method that couldn't be called from anywhere _except_ a constructor would be unnecessarily restrictive.

Now I understand that such methods might want to set the values directly and that `init` could be a way to allow them to do so. Personally I would find it confusing and would just refactor so that they returned the values rather than assigning them (but to each his own). The one "downside" is that you'd have to make sure all the constructors chained to each other so you weren't duplicating the method call/assignments in multiple places. I suppose that's not always possible and `init` methods could allow you to consolidate in those cases. Still I'd find that a bit confusing and would opt for the longer form, even if it meant some duplication.
